htlcswitch/switch.go

```diff
@@ -406,30 +424,46 @@ func (s *Switch) handleLocalDispatch(payment *pendingPayment, packet *htlcPacket
 		// Notify the user that his payment was successfully proceed.
 		payment.err <- nil
 		payment.preimage <- htlc.PaymentPreimage
-		s.removePendingPayment(payment.amount, payment.paymentHash)
+		s.removePendingPayment(packet.incomingHTLCID)
 
 	// We've just received a fail update which means we can finalize the
 	// user payment and return fail response.
 	case *lnwire.UpdateFailHTLC:
-		// We'll attempt to fully decrypt the onion encrypted error. If
-		// we're unable to then we'll bail early.
-		failure, err := payment.deobfuscator.DecryptError(htlc.Reason)
-		if err != nil {
-			userErr := fmt.Sprintf("unable to de-obfuscate "+
-				"onion failure, htlc with hash(%x): %v",
-				payment.paymentHash[:], err)
-			log.Error(userErr)
-			payment.err <- &ForwardingError{
+		var failure *ForwardingError
+		if packet.localFailure {
+			var userErr string
+			r := bytes.NewReader(htlc.Reason)
+			failureMsg, err := lnwire.DecodeFailure(r, 0)
+			if err != nil {
+				userErr = fmt.Sprintf("unable to de-obfuscate onion failure, "+
```

Comment:
Error message is now incorrect. It should state it's unable to _decode_ the wire error.